[Java Client] Optimize batch flush scheduling

michaeljmarshall opened the pull request:
Master Issue: https://github.com/apache/pulsar/issues/11100

### Motivation

As observed in https://github.com/apache/pulsar/issues/11100, the Java client producer consumes cpu even when doing nothing. This is especially true when using many producers. Instead, we can make it so that the batch timer is only scheduled when it has messages to deliver or just fired and delivered messages.

If there are concerns about this optimization, I will need to take some time to complete benchmarks.

### Modifications

* Skip message batch delivery if the producer is not in `Ready` state. As a consequence, ensure that messages pending in the `batchMessageContainer` are still eligible to fail for `sendTimeout`.
* If there is no current batch flush task, schedule a single flush task when a message is added to a batch message container (assuming the message does not also trigger the delivery of the batch and the producer is in READY state).
* Schedule another batch flush task if and only if the batch flush task itself was responsible for sending messages.
* Keep track of `lastBatchSendNanoTime`, and only flush a batch message container if the `BatchingMaxPublishDelayMicros` time has passed since the last send time.
* Note that the timer task is only ever updated within `synchronized (this)` block, so we are guaranteed to have a consistent view free of race conditions. (There is one race condition, and that is that an existing batch timer might get canceled while it is starting. This is of little consequence since it'll result in either no delivery or a small batch.)

### Verifying this change

There are existing tests that cover batch message delivery. Specifically, the `BatchMessageTest` in the `pulsar-broker` module covers these changes.

### Does this pull request potentially affect one of the following parts:

This is a backwards compatible change.

### Documentation

- [x] `no-need-doc`

lhotari commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:
```diff
@@ -1947,8 +1933,34 @@ protected void triggerFlush() {
+    private void maybeTriggerBatchTimerTask() {
+        if (this.batchTimerTask == null) {
+            this.batchTimerTask = cnx().ctx().executor()
+                    .schedule(catchingAndLoggingThrowables(() -> {
```
This could be refactored further. A separate method would be helpful to reduce the indention levels.

michaeljmarshall replied:
I agree.

merlimat replied:
Also returning early could help: 


```java
   if (this.batchTimerTask != null) {
       return;
   }
```

lhotari commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:
```diff
@@ -148,6 +148,11 @@
 
     private final ConnectionHandler connectionHandler;
 
+    // A batch timer task is only ever scheduled when a message is added to a message batch without also triggering a
```
this is no more a "timer task". 
[javadoc](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ScheduledExecutorService.html#schedule(java.lang.Runnable,long,java.util.concurrent.TimeUnit))
schedule method returns "a ScheduledFuture representing pending completion of the task and whose get() method will return null upon completion"
scheduleAtFixedDelay returned "a ScheduledFuture representing pending completion of the series of repeated tasks. The future's [get()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/Future.html#get()) method will never return normally, and will throw an exception upon task cancellation or abnormal termination of a task execution."

For this reason, it would be better to rename `batchTimerTask` field accordingly. It could be simply `batchTask`.

michaeljmarshall replied:
Good point. I think I'll rename it to `batchFlushTask`.

lhotari commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:
```diff
@@ -1721,10 +1705,12 @@ private void closeProducerTasks() {
-        if (batchTimerTask != null) {
-            batchTimerTask.cancel(false);
-            this.batchTimerTask = null;
+        synchronized (this) {
+            ScheduledFuture<?> batchTimerTask = this.batchTimerTask;
+            if (batchTimerTask != null) {
+                batchTimerTask.cancel(false);
+                this.batchTimerTask = null;
+            }
         }
```
Since this is no more a "timer task", it would be better to let the task complete gracefully. It needs some timeout when waiting for the completion.

michaeljmarshall replied:
Are you suggesting we remove this from the close block? I'm fine with that since it will already exit early based on the producer's state.

lhotari replied:
Yes, removing it from the close block or waiting for the current task to complete. Why would you cancel it? What is the desired behavior in cancelling?

lhotari commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:
```diff
@@ -1947,8 +1933,34 @@ protected void triggerFlush() {
+
     // must acquire semaphore before enqueuing
     private void batchMessageAndSend() {
+        if (this.batchTimerTask != null) {
+            // Cancel batch timer task since we're sending the batch now. (Don't interrupt; it could be this future.)
+            this.batchTimerTask.cancel(false);
+            this.batchTimerTask = null;
+        }
```
The previous in-flight batch task shouldn't be cancelled. That would cause problems. There can be multiple tasks queued and in flight at once. It's a valid case that the previous task might not be completed when the next batch task is created.

michaeljmarshall replied:
My proposed design is to have either one or no tasks in-flight. If we don't cancel tasks, we'll end up with a task per batch message, which will result in smaller batches and increased CPU utilization when sending many batches.

What problems do you see cancellation causing?

This design relies on the fact that the batch task is only ever created or canceled from within a lock on `this` producer. There are no races. When we are flushing a batch, there are no pending batch messages. The next time the lock on `this` is acquired and a message is enqueued, it can correctly schedule the next batch task.

I admit that cancellation adds CPU overhead. I haven't bench marked this to compare it to the current design.

merlimat replied:
Yes, there should only be 1 batch being packed at once, because we're reusing some of the containers

lhotari replied:
> My proposed design is to have either one or no tasks in-flight. If we don't cancel tasks, we'll end up with a task per batch message, which will result in smaller batches and increased CPU utilization when sending many batches.

I don't see how this could work in the current solution. The chance for a data race remains: it's possible that the scheduled task has started execution and runs batchMessageAndSend concurrently with the batchMessageAndSend that results from a batch being filled up. Cancelling the previous task doesn't resolve this issue in the given example. 
batchMessageAndSend should be run within a synchronized block so it takes care of races at that level. However, in theory that this happens:
1. batchMessageAndSend gets triggered since it's full
2. a new message is added to batch
3. batchMessageAndSend gets triggered by the timer
In the above example, it's possible that the batchMessageAndSend gets sent off with just 1 message in the batch message.

> What problems do you see cancellation causing?

It doesn't prevent the data race explained in the given example. If the execution hasn't yet started, it will not run the task, but there's a chance for races after the point when the task is already executing.

michaeljmarshall replied:
I should add that there is an issue with the earlier logic where it wouldn't have worked with a producer that had to reconnect. I just pushed up what I believe are correct changes. I will add a test later today to verify the reconnect case. The other cases are already covered in the the `BatchMessageTest`.

michaeljmarshall replied:
@lhotari - I updated the logic to prevent early batch flushes. Please take a look.

merlimat commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:
```diff
@@ -1947,8 +1933,34 @@ protected void triggerFlush() {
+                        }
+                        // No need to prematurely cut batches if we're not ready to send them.
+                        if (getState() != State.Ready) {
+                            return;
```
I think we need to ensure that we schedule or trigger `batchMessageAndSend()` once the producer gets reconnected, otherwise it might not get re-triggered until new `send()` are issued. It could leave the existing incomplete batch for indefinite time.

michaeljmarshall replied:
Great point, thanks for catching this case.

michaeljmarshall replied:
I handle this case now. I still need to add a test to cover this case, though.

lhotari commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:
```diff
@@ -1947,8 +1925,38 @@ protected void triggerFlush() {
     // must acquire semaphore before enqueuing
     private void batchMessageAndSend() {
+        if (this.batchFlushTask != null) {
+            // Cancel batch timer task since we're sending the batch now.
```
this comment references the previous concept "timer task"

michaeljmarshall commented:
@lhotari and @merlimat - I realized that the cost of canceling and rescheduling a batch message timer for every batch might not be ideal, especially in high throughput producers. Therefore, I switched the logic to match that of the `sendTimeout`, which essentially reschedules a timeout if an insufficient amount of time has passed since an event (in this PR's case, that is the time since last sending a message).

I also rand into an issue with the `SimpleProducerConsumerStatTest#testSendTimeout`. It exposed the fact that my design did not allow for queued batch messages to get timed out when the producer is disconnected. I updated the logic with a naive solution. The one "surprising" affect of my solution might be that a message might get up to double the send timeout if the producer reconnects at the "right" time. I think this is fine, but let me know if this tradeoff is a problem.

lhotari reviewed (approved):
Good work @michaeljmarshall !
